beginner_source/Pretraining_Vgg_from_scratch.py

```diff
@@ -0,0 +1,665 @@
+        img=np.array(img,dtype=np.float32)
+
+
+        if self.transform is not None:
+            img = self.transform(image=img)
+            if len(img['image'].shape) == 3 and self.val==False :
+                img = A.RGBShift()(image=img['image'])
+            img = img['image']
+
+        if self.target_transform is not None:
+            target = self.target_transform(target)
+
+        img=img.transpose((2,0,1))
+
+        return img, target
+
+
+######################################################################
+# Define Model
+# ------------
+# 
+
+
+######################################################################
+# | In VGG paper, they do experiment over 6 models. model A is 11 layers,
+#   model B is 13 layers, model C is 16 layers , model D is 16 layers and
+#   model D is 19 layers . you can train all version of models to
+#   reproduce VGG .
+# | ``Config_Channels`` means output channels and ``Config_kernels`` means
+#   kernel size.
+# 
+
+import torch
+from torch import nn
+
+
+Config_channels = {
+"A" : [64,"M" , 128,  "M"  , 256,256,"M" ,512,512 ,"M" , 512,512,"M"] ,
+"A_lrn" : [64,"LRN","M" , 128,  "M"  , 256,256,"M" ,512,512 ,"M" , 512,512,"M"] ,
+"B" :[64,64,"M" , 128,128,  "M"  , 256,256,"M" ,512,512 ,"M" , 512,512,"M"]  ,
+"C" : [64,64,"M" , 128,128,  "M"  , 256,256,256,"M" ,512,512 ,512,"M" , 512,512,512,"M"] ,
+"D" :[64,64,"M" , 128,128,  "M"  , 256,256,256,"M" ,512,512 ,512,"M" , 512,512,512,"M"] ,
+"E" :[64,64,"M" , 128,128,  "M"  , 256,256,256,256,"M" ,512,512 ,512,512,"M" , 512,512,512,512,"M"]         ,
+
+}
+
+
+
+Config_kernel = {
+"A" : [3,2 , 3,  2  , 3,3,2 ,3,3 ,2 , 3,3,2] ,
+"A_lrn" : [3,2,2 , 3,  2  , 3,3,2 ,3,3 ,2 , 3,3,2] ,
+"B" :[3,3,2 , 3,3,  2  , 3,3,2 ,3,3 ,2 , 3,3,2]  ,
+"C" : [3,3,2 , 3,3,  2  , 3,3,1,2 ,3,3 ,1,2 , 3,3,1,2] ,
+"D" :[3,3,2 , 3,3,  2  , 3,3,3,2 ,3,3 ,3,2 , 3,3,3,2] ,
+"E" :[3,3,2 , 3,3,  2  , 3,3,3,3,2 ,3,3 ,3,3,2 , 3,3,3,3,2]         ,
+
+}
+
+
+######################################################################
+# We define model class that generate model in choice of 6 versions.
+# 
+
+def make_feature_extractor(cfg_c,cfg_k):
+    feature_extract = []
+    in_channels = 3
+    i = 1
+    for  out_channels , kernel in zip(cfg_c,cfg_k) :
+        # print(f"{i} th layer {out_channels} processing")
+        if out_channels == "M" :
+            feature_extract += [nn.MaxPool2d(kernel,2) ]
+        elif out_channels == "LRN":
+            feature_extract += [nn.LocalResponseNorm(5,k=2) , nn.ReLU()]
+        elif out_channels == 1:
+            feature_extract+= [nn.Conv2d(in_channels,out_channels,kernel,stride = 1) , nn.ReLU()]
+        else :
+            feature_extract+= [nn.Conv2d(in_channels,out_channels,kernel,stride = 1 , padding = 1) , nn.ReLU()]
+
+        if isinstance(out_channels,int) :   in_channels = out_channels
+        i+=1
+    return nn.Sequential(*feature_extract)
+
+
+class Model_vgg(nn.Module) :
+    def __init__(self,version , num_classes):
```

Comment:
style nit: more flexibly than passing `version` here, the model could take in the config channels / config kernel lists directly (or config could be a first-class object for clarity) and the mapping between model variant (e.g. "A") -> config data can be referenced when constructing the model. This allows the model to support additional variants outside of those in the VGG paper for experimentation.

```python
# configs contains mappings like "A" -> first-class Config object
mod = Model_vgg(configs["A"], NUM_CLASSES)
```